### PR TITLE
Retain original appointment with recurring pattern

### PR DESCRIPTION
**Pull request by venkatavineela**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentRecurringPatternServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    private void setAppointmentAuditForStatusChange(Appointment appointment, Date onDate) {
+        String notes = onDate != null ? onDate.toInstant().toString() : null;
+        updateAppointmentAudits(appointment, notes);
+    }
+
+    private void updateAppointmentAudits(Appointment appointment, String notes) {
```
Comment (binduak) [inline api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentRecurringPatternServiceImpl.java:149]: this "updateAppointmentAudits" method has implementations in AppointmentRecurringPatternServiceImpl, AppointmentRecurringPatternServiceImpl and SingleAppointmentRecurringPatternMapper

File: omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java [hunk 2]
```diff
@@ -237,23 +259,42 @@ public AppointmentDefaultResponse getAppointmentByUuid(@RequestParam(value = "uu
...
     public ResponseEntity<Object> editAppointment(@Valid @RequestBody AppointmentRequest appointmentRequest) {
         try {
-            boolean applyForAll = appointmentRequest.getApplyForAll() == null
-                    ? false : appointmentRequest.getApplyForAll();
+            final boolean isValidAppointmentRequest = appointmentRequestEditValidator.validate(appointmentRequest);
+            if (!isValidAppointmentRequest)
+                throw new APIException(appointmentRequestEditValidator.getError());
+            boolean applyForAll = appointmentRequest.requiresUpdateOfAllRecurringAppointments();
+            final String appointmentRequestUuid = appointmentRequest.getUuid();
             if (applyForAll) {
```
Comment (binduak) [inline omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java:267]: we are not using clientTimeZone variable here. As part of the code refactoring i see we have already fetching  **appointmentRequest.getTimeZone();** in "AllAppointmentRecurringPatternMapper" class. I think we can handle this Timezone exception there itself.

File: omod/src/main/java/org/openmrs/module/appointments/web/mapper/SingleAppointmentRecurringPatternMapper.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Autowired
+    private AppointmentServiceHelper appointmentServiceHelper;
+
+    @Override
+    public AppointmentRecurringPattern fromRequest(AppointmentRequest appointmentRequest){
+        String uuid = appointmentRequest.getUuid();
+        Appointment appointment = appointmentsService.getAppointmentByUuid(uuid);
+        Appointment newAppointment;
+        if(appointment==null)
```
Comment (binduak) [inline omod/src/main/java/org/openmrs/module/appointments/web/mapper/SingleAppointmentRecurringPatternMapper.java]: Can we use AppointmentValidator class here to validate the appointment ??

**Review by binduak — CHANGES_REQUESTED** [review]

**Review by bsneha90 — COMMENTED** [review]
For single appointment mapper, we need to run and make sure the related appointment is getting created correctly for first time edit an next edit.

File: api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentRecurringPatternServiceImpl.java (hunk 1 above)
Comment (venkatavineela) [reply api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentRecurringPatternServiceImpl.java:149]: We have to move some part of code from api to omod. We can play a separate card for that where we can refactor this.

File: omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java (hunk 2 above)
Comment (venkatavineela) [reply omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java:267]: Moved this validation to "AllAppointmentRecurringPatternMapper"

File: omod/src/main/java/org/openmrs/module/appointments/web/mapper/SingleAppointmentRecurringPatternMapper.java (hunk 3 above)
Comment (venkatavineela) [reply omod/src/main/java/org/openmrs/module/appointments/web/mapper/SingleAppointmentRecurringPatternMapper.java]: Removed this part as we are doing this as part of editAppointmentRequestValidation.

**Comment by binduak** [comment]
@bsneha90 We have tested the issue that you have mentioned for PR.  Related appointment is getting created correctly for first time edit an next edit.

**Pull request merged**